### PR TITLE
GraphQl: parse input&messenger metadata

### PR DESCRIPTION
**Pull request by soyuka**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | na
| License       | MIT
| Doc PR        | na

We forgot to take graphql mutations into consideration when retrieving the metadata. This now works correctly when performing a `createBookmark` mutation on my resource:

```
/**
 * @ApiResource(
 *      graphql={"create"={"input"=BookmarkInput::class, "messenger"="input"}}
 * )
 */
class Bookmark {}
``` 

It'll forward `BookmarkInput` to the messenger via the mutation :fire: :fire:  :fire:  . 

Note: for naming consistency I took the liberty of changing `graphql_operation_name` to `graphql_mutation_name` in our internal `$context` because else it's confusing.

### CODE REVIEW

File: features/main/input_output.feature [hunk 1]
```diff
@@ -287,3 +287,30 @@ Feature: DTO input and output
       "data": 123
     }
     """
+
+  @!mongodb
```
Comment (alanpoulain) [inline features/main/input_output.feature]: Why?

File: features/main/input_output.feature [hunk 2]
```diff
@@ -287,3 +287,30 @@ Feature: DTO input and output
...
+  Scenario: Use messenger with graphql and an input where the handler gives a synchronous result
+    When I send the following GraphQL request:
+    """
+    mutation {
+      createMessengerWithInput(input: {var: "test"}) {
+        clientMutationId, messengerWithInput { id, name }
```
Comment (alanpoulain) [inline features/main/input_output.feature]: Line break?

File: features/main/input_output.feature [hunk 3]
```diff
@@ -287,3 +287,30 @@ Feature: DTO input and output
...
+    }
+    """
+    Then the response status code should be 200
+    And the response should be in JSON
+    And the JSON should be equal to:
+    """
+    {
+      "data": {
+        "createMessengerWithInput": {
+          "clientMutationId": null,
```
Comment (alanpoulain) [inline features/main/input_output.feature]: You can remove it.

File: src/GraphQl/Resolver/Factory/ItemMutationResolverFactory.php [hunk 4]
```diff
@@ -65,9 +65,9 @@ public function __construct(IriConverterInterface $iriConverter, DataPersisterIn
         $this->validator = $validator;
     }
 
-    public function __invoke(string $resourceClass = null, string $rootClass = null, string $operationName = null): callable
```
Comment (dunglas) [inline src/GraphQl/Resolver/Factory/ItemMutationResolverFactory.php]: I would revert this change. This is triggered for queries to, and queries aren't mutations. Mutations + Queries = operations.

File: src/GraphQl/Type/SchemaBuilder.php [hunk 5]
```diff
@@ -419,7 +419,7 @@ private function getResourceObjectType(?string $resourceClass, ResourceMetadata
             return $this->graphqlTypes[$shortName];
         }
 
-        $ioMetadata = $resourceMetadata->getAttribute($input ? 'input' : 'output');
+        $ioMetadata = null === $mutationName ? $resourceMetadata->getAttribute($input ? 'input' : 'output') : $resourceMetadata->getGraphqlAttribute($mutationName, $input ? 'input' : 'output', null, true);
```
Comment (alanpoulain) [inline src/GraphQl/Type/SchemaBuilder.php]: Can't we have a specific output for a query?

File: src/GraphQl/Type/SchemaBuilder.php [hunk 6]
```diff
@@ -492,7 +492,7 @@ private function getResourceObjectTypeFields(?string $resourceClass, ResourceMet
 
         if (null !== $resourceClass) {
             foreach ($this->propertyNameCollectionFactory->create($resourceClass) as $property) {
-                $propertyMetadata = $this->propertyMetadataFactory->create($resourceClass, $property, ['graphql_operation_name' => $mutationName ?? 'query']);
+                $propertyMetadata = $this->propertyMetadataFactory->create($resourceClass, $property, ['graphql_mutation_name' => $mutationName ?? 'query']);
```
Comment (alanpoulain) [inline src/GraphQl/Type/SchemaBuilder.php]: It seems strange to use `graphql_mutation_name` when it's a query. And there will be custom queries too.

File: features/main/input_output.feature (hunk 1 above)
Comment (soyuka) [reply features/main/input_output.feature]: Because I don't want to create a document for this and also because this entity isn't even tight to a database persister so there's no point of running it with mongodb.

Comment (alanpoulain) [reply features/main/input_output.feature]: Yes but I think it would be better to have it working with MongoDB anyway. This tag is only here when the functionality is not compatible with MongoDB.
